caas/kubernetes/provider/specs/v2_test.go

```diff
@@ -0,0 +1,591 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package specs_test
+
+import (
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+	core "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	k8sspecs "github.com/juju/juju/caas/kubernetes/provider/specs"
+	"github.com/juju/juju/caas/specs"
+	"github.com/juju/juju/testing"
+)
+
+type v2SpecsSuite struct {
+	testing.BaseSuite
+}
+
+var _ = gc.Suite(&v2SpecsSuite{})
+
+var versionHeader = `
+version: 2
+`[1:]
+
+func (s *v2SpecsSuite) TestParse(c *gc.C) {
+
+	specStrBase := versionHeader + `
+omitServiceFrontend: true
+containers:
+  - name: gitlab
+    image: gitlab/latest
+    command:
+      - sh
+      - -c
+      - |
+        set -ex
+        echo "do some stuff here for gitlab container"
+    args: ["doIt", "--debug"]
+    workingDir: "/path/to/here"
+    ports:
+      - containerPort: 80
+        name: fred
+        protocol: TCP
+      - containerPort: 443
+        name: mary
+    kubernetes:
+      imagePullPolicy: Always
+      securityContext:
+        runAsNonRoot: true
+        privileged: true
+      livenessProbe:
+        initialDelaySeconds: 10
+        httpGet:
+          path: /ping
+          port: 8080
+      readinessProbe:
+        initialDelaySeconds: 10
+        httpGet:
+          path: /pingReady
+          port: www
+    config:
+      attr: foo=bar; name['fred']='blogs';
+      foo: bar
+      restricted: 'yes'
+      switch: on
+    files:
+      - name: configuration
+        mountPath: /var/lib/foo
+        files:
+          file1: |
+            [config]
+            foo: bar
+  - name: gitlab-helper
+    image: gitlab-helper/latest
+    ports:
+    - containerPort: 8080
+      protocol: TCP
+  - name: secret-image-user
+    imageDetails:
+        imagePath: staging.registry.org/testing/testing-image@sha256:deed-beef
+        username: docker-registry
+        password: hunter2
+  - name: just-image-details
+    imageDetails:
+        imagePath: testing/no-secrets-needed@sha256:deed-beef
+  - name: gitlab-init
+    image: gitlab-init/latest
+    init: true
+    kubernetes:
+      imagePullPolicy: Always
+    command:
+      - sh
+      - -c
+      - |
+        set -ex
+        echo "do some stuff here for gitlab-init container"
+    args: ["doIt", "--debug"]
+    workingDir: "/path/to/here"
+    ports:
+    - containerPort: 80
+      name: fred
+      protocol: TCP
+    - containerPort: 443
+      name: mary
+    config:
+      foo: bar
+      restricted: 'yes'
+      switch: on
+configMaps:
+  mydata:
+    foo: bar
+    hello: world
+service:
+  annotations:
+    foo: bar
+pod:
```

Comment:
I'm thinking perhaps this "pod" section belongs in the kubernetes.yaml file.
Things like restartPolicy could be generic, but we say securtityContext is k8s specific in the containers section so logically it should be k8s specific here too? I think it's easier to start out with stuff being k8s specific and move across as we determine how to model